### ISSUE TITLE
Monitoring: label of type INT64 rejected

### ISSUE THREAD

**Issue by kenashcraft** [GoogleCloudPlatform/google-cloud-python#4825]
I'm trying to write a simple HTTP prober and write the resulting status code to StackDriver.  It doesn't seem to like labels of type INT64

```
Traceback (most recent call last):
  File "./prober.py", line 44, in <module>
    client.write_point(metric, resource, 1)
  File "/Users/ken/code/venv/lib/python3.6/site-packages/google/cloud/monitoring/client.py", line 601, in write_point
    self.write_time_series([timeseries])
  File "/Users/ken/code/venv/lib/python3.6/site-packages/google/cloud/monitoring/client.py", line 546, in write_time_series
    data={'timeSeries': timeseries_dict})
  File "/Users/ken/code/venv/lib/python3.6/site-packages/google/cloud/_http.py", line 293, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.BadRequest: 400 POST https://monitoring.googleapis.com/v3/projects/<projectIdRedacted>/timeSeries/: Invalid value at 'time_series[0].metric.labels[0].value' (TYPE_STRING), 200
```
It would be helpful if the error message said which label had an invalid value.  After some process of elimination, I confirmed that it is the 'response_code' label.

Here is my code:
```
from http.client import HTTPConnection

from google.cloud.monitoring import (
    Client,
    LabelDescriptor, 
    LabelValueType,
    MetricKind,
    ValueType
)

client = Client()

response_code_label = LabelDescriptor('response_code', LabelValueType.INT64,
                                      description='HTTP status code')
path_label = LabelDescriptor('path', LabelValueType.STRING,
                             description='The requested path (e.g. "/index.html")')
response_descriptor = client.metric_descriptor(
    'custom.googleapis.com/client_ui/response_count',
    metric_kind=MetricKind.CUMULATIVE,
    value_type=ValueType.INT64,
    labels=[response_code_label, path_label],
    description='Cumulative count of HTTP responses.'
)
response_descriptor.create()

host = 'www.google.com'
path = '/'

conn = HTTPConnection(host, timeout=2)
conn.request('GET', path)
response = conn.getresponse()
print('{} {}'.format(response.status, response.reason))

# https://cloud.google.com/monitoring/custom-metrics/creating-metrics#which-resource
resource = client.resource(type_='global', labels={})
metric = client.metric(
    type_='custom.googleapis.com/client_ui/response_count',
    labels={
        'response_code': response.status,
        'path': path,
    }
)
client.write_point(metric, resource, 1)
```

It seems to work if I change the type of the `response_code` to STRING:
```
response_code_label = LabelDescriptor('response_code', LabelValueType.STRING,
                                      description='HTTP status code')
...  # Same as above.
metric = client.metric(
    type_='custom.googleapis.com/client_ui',
    labels={
        'response_code': str(response.status),
        'path': path,
    }
)
client.write_point(metric, resource, 1)
```

- OS type and version: Mac OS 10.13.3
- Python version and virtual environment information `python --version`: 3.6.0
- google-cloud-python version `pip show google-cloud`, `pip show google-<service>` or `pip freeze`

$ pip show google-cloud-monitoring
Name: google-cloud-monitoring
Version: 0.28.0
Summary: Python Client for Stackdriver Monitoring
Home-page: https://github.com/GoogleCloudPlatform/google-cloud-python
Author: Google Cloud Platform
Author-email: googleapis-publisher@google.com
License: Apache 2.0
Location: /Users/ken/code/venv/lib/python3.6/site-packages
Requires: google-api-core, google-cloud-core

**Comment by jonparrott** [comment]
@liyanhui1228 can you help here?

**Comment by liyanhui1228** [comment]
@jonparrott OK, taking a look at it.

**Comment by liyanhui1228** [comment]
@kenashcraft Thanks for reporting! This is actually an intended behavior according to [metric.proto](https://github.com/googleapis/googleapis/blob/70402cea3d4f76878ebe83907c4ccc1c604b39b7/google/api/metric.proto#L195). The `labels` field of `Metric` instance only accepts string type values. While the value type of the `LabelDescriptor` is for the backend processing of the label values, like sum, average or count. The documentation is somewhat confusing, we need to add a comment to tell customers that the `labels` field in metric can only be `str` when creating a `Metric` instance.

**Comment by kenashcraft** [comment]
@liyanhui1228 Hmm... that doesn't seem right to me.  Why have `LabelDescriptor` support multiple types if only string is allowed?  Why have `LabelDescriptor` at all in that case.

The line you're referring to is
```
map<string, string> labels = 2;
```
I wonder if the `string` value was simply a matter of convenience because you can't put a union ("int or string or bool") in the value for a map in protocol buffers.  Perhaps the intention was that clients should encode their value in a string, and the server will decode it back to the proper type.

Including @wora who appears to be the original author.

**Comment by wora** [comment]
LabelDescriptor defines the semantic type of the value. The reason to always use string as wire format is to simplify user experience and make the code much more efficient.

For example, you can try to copy a union to another union, you will see how many lines of code you have to write. Another example, if two labels have different value types (by mistake or other reason), copying value would be a very bad experience.

**Comment by kenashcraft** [comment]
@wora Right, string is used as the wire format, but the semantics are important.  So the client should use the LabelDescriptor to say that the field is, say, an INT64.  It seems like the google-cloud-python library should automatically convert all label values to string before sending them to the server.  The ideal code looks like this:
```
response_code_label = LabelDescriptor('response_code', LabelValueType.INT64,
                                      description='HTTP status code')
...  # Same as above.
metric = client.metric(
    type_='custom.googleapis.com/client_ui',
    labels={
        'response_code': response.status,  # This is an int -- client.metric() converts it to a string.
        'path': path,
    }
)
client.write_point(metric, resource, 1)
```

**Comment by tseaver** [comment]
@jonparrott This looks like a feature request to me, rather than a bug (and certainly not a P1, as the user has a clear workaround).

**Comment by jonparrott** [comment]
Agreed. Always feel free to modify labels as information and situations change.

**Comment by jonparrott** [comment]
@wora @liyanhui1228 can you confirm whether this is a feature we want to add?

**Comment by liyanhui1228** [comment]
@jonparrott Yes adding this feature SGTM.

**Comment by chemelnucfin** [comment]
Closed via #4910
